### DISCUSSION TITLE
Announcing vf2: A subgraph isomorphism algorithm in Rust

### DISCUSSION THREAD

**Discussion by OwenTrokeBillard** [OwenTrokeBillard/vf2#1] [Announcements]
<div align="center">

<img src="https://raw.githubusercontent.com/OwenTrokeBillard/vf2/main/images/icon.svg" width="128" alt="icon"/>

# `vf2`

### VF2 subgraph isomorphism algorithm in Rust

[![GitHub Repo stars](https://img.shields.io/github/stars/OwenTrokeBillard/vf2)](https://github.com/OwenTrokeBillard/vf2)
[![crates.io](https://img.shields.io/crates/v/vf2.svg)](https://crates.io/crates/vf2)
[![docs.rs](https://img.shields.io/docsrs/vf2)](https://docs.rs/vf2/latest/vf2)

<br/>
</div>

I am pleased to announce the release of `vf2` 1.0.0 on crates.io. This crate implements the VF2 subgraph isomorphism algorithm [1]. It can find [graph isomorphisms](https://en.wikipedia.org/wiki/Graph_isomorphism), [subgraph isomorphisms](https://en.wikipedia.org/wiki/Subgraph_isomorphism_problem), and [induced subgraph isomorphisms](https://en.wikipedia.org/wiki/Induced_subgraph_isomorphism_problem). Graphs can be directed or undirected.

Please [leave a star](https://github.com/OwenTrokeBillard/vf2) and provide feedback on how this crate can be improved.

# Motivation

Subgraph matching is the task of finding instances of a query graph within a larger data graph. It is useful when
searching for patterns in a graph structure, such as relationships in a social network. It is a fundamental problem in
graph theory with applications in pattern recognition, databases, security, and biochemistry.

Consider a network like LinkedIn where each node is a person, and each edge represents a connection. You are tasked with
finding cases where five software developers and a doctor all know each other. You can make a query graph with
developers and a doctor, and search for instances of that query in the larger network.

# What is a subgraph isomorphism?

A graph is a structure consisting of a set of objects where some pairs of objects are connected. A graph isomorphism is
a one-to-one correspondence between two graphs such that objects connected in one are also connected in the other.

### Graph isomorphism

For two graphs to be isomorphic, there must be a one-to-one correspondence between nodes such that neighbors in one are
also neighbors in the other. The query and data graphs in the following image are isomorphic.

![graph-isomorphism.svg](https://raw.githubusercontent.com/OwenTrokeBillard/vf2/main/images/graph-isomorphism.svg)

### Subgraph isomorphism

It is often desirable to find instances of one graph within another. To do this, we search for subgraph isomorphisms. A
subgraph isomorphism is when one graph is isomorphic to a subgraph of another. There are two subgraph isomorphisms in
the following image.

![subgraph-isomorphism.svg](https://raw.githubusercontent.com/OwenTrokeBillard/vf2/main/images/subgraph-isomorphism.svg)

### Induced subgraph isomorphism

An induced subgraph isomorphism is the same as a subgraph isomorphism except that the subgraph must be induced. Edges in
the data subgraph must also exist in the query graph.

![induced-subgraph-isomorphism.svg](https://raw.githubusercontent.com/OwenTrokeBillard/vf2/main/images/induced-subgraph-isomorphism.svg)

# Example

This example shows how to find subgraph isomorphisms. Please see the repository for usage instructions.

```rust
use petgraph::data::{Element, FromElements};
use petgraph::graph::DiGraph;

fn main() {
    // Create query graph.
    let query = DiGraph::<i32, ()>::from_elements([
        Element::Node { weight: 0 },
        Element::Node { weight: 1 },
        Element::Edge { source: 0, target: 1, weight: () },
    ]);

    // Create data graph.
    let data = DiGraph::<i32, ()>::from_elements([
        Element::Node { weight: 0 },
        Element::Node { weight: 1 },
        Element::Node { weight: 2 },
        Element::Edge { source: 0, target: 1, weight: () },
        Element::Edge { source: 1, target: 2, weight: () },
    ]);

    // Find subgraph isomorphisms.
    let isomorphisms = vf2::subgraph_isomorphisms(&query, &data).vec();
    assert_eq!(isomorphisms, vec![vec![0, 1], vec![1, 2]]);
}
```

# References

[1] L. P. Cordella, P. Foggia, C. Sansone, and M. Vento, “A (sub)graph isomorphism algorithm for matching large graphs,” IEEE Transactions on Pattern Analysis and Machine Intelligence, vol. 26, no. 10, pp. 1367–1372, Oct. 2004, doi: https://doi.org/10.1109/tpami.2004.75.

[2] A. Jüttner and P. Madarasi, “VF2++—An improved subgraph isomorphism algorithm,” Discrete Applied Mathematics, vol. 242, pp. 69–81, Jun. 2018, doi: https://doi.org/10.1016/j.dam.2018.02.018.

**Comment by ThijsVromen** [comment]
Really cool!  I'm currently implementing the Weisfeiler-Leman algorithm (an incomplete heuristic for graph isomorphism testing/neighbourhood hashing). Similar to your work I'm using petgraph as a foundation, but most likely won't be able to integrate it in that crate any time soon because the maintainers there are very busy, so will probably (first) create a standalone crate for it. The way you (made and) present your crate is definitely an inspiration for that crate! 

I'd love to hear if you have any advice on for example pitfalls/convenient things you came across while building this. I'm already particularly curious why you decided to make a `Graph` trait yourself, and why you decided to implement that (only) for the `Graph` type from petgraph. Currently, I'm wrestling a bit with deciding for what types of petgraph graphs to implement my algorithms (and how to make it generic over these types).

**Reply by OwenTrokeBillard** [reply]
@Anemoonvis A heuristic approach sounds promising. I look forward to seeing your work.

One challenge I encountered was expressing the search as an iterator. Iterators have a few advantages:
- **Memory Efficiency:** Results can be calculated, inspected, and discarded one by one, minimizing memory consumption.
- **Early Stopping:** Users can stop the search early if desired results are found or if computational resources are limited.
- **Flexibility:** Results can be collected into a vector or processed individually as needed.

However, to leverage iterators, the search process must be resumable. This requires storing state information between steps, which increases complexity. A recursive function that populates a vector is a simpler approach. It's up to you if the extra complexity is worth it.

The `Graph` trait provides flexibility for users to integrate `vf2` with their preferred graph library. While `petgraph` is used as a default implementation, the crate allows users to disable the `petgraph` feature and implement the `Graph` trait for their own graph types.

**Reply by ThijsVromen** [reply]
Thanks for the response! 

> The Graph trait provides flexibility for users to integrate vf2 with their preferred graph library. While petgraph is used as a default implementation, the crate allows users to disable the petgraph feature and implement the Graph trait for their own graph types.

That makes a lot of sense!

Sounds really interesting that you turned it into an iterator. I'm not sure how applicable it would be for the WL algorithm, since that is not search algorithm but a refinement algorithm; each iteration has to consider the graph as a whole anyways. So, for now, I just have an option to limit the number of iterations (which is also how e.g. NetworkX in python does it) or to let it stop once the refinement has stabilised (that is how it is used in the literature). Frankly, I also think making it an iterator could also be a bit hard for me at this point. This is the first Rust code I'm writing besides Rustlings/Rustbook exercises, haha. 

Did you come up with the inputs/outputs for the isomorphism tests yourself, or did you get them from somewhere? At some point I'll need a bunch of examples for verification and benchmarking, but not yet sure where to find a nice dataset. (This is part of my bachelor thesis project, so I'll be using those benchmarks to make some nice graphs to talk about in there, I can image you wouldn't necessarily put that effort in for a project like this normally)

**Reply by OwenTrokeBillard** [reply]
I came up with the test cases myself. You are welcome to use them. I searched for a public dataset but didn't find any either.

I'm glad to hear you're using Rust for your thesis project.

**Reply by ThijsVromen** [reply]
Thanks, that seems like a good start!
